### PR TITLE
[12.0] Backport of 13.0 multiple improvements

### PR DESCRIPTION
**Pull request by etobella**
Backport of #305 

@simahawk @guewen

### CODE REVIEW

**Comment by etobella** [comment]
Also added #312 #310 and #309

**Review by simahawk — APPROVED** [review]
just code review, LG

**Review by guewen — APPROVED** [review]
Thanks a lot!

**Comment by guewen** [comment]
Hi @sbidoul, @gurneyalex,

Runbot fails because of this on the PRs for 12.0:

```
Unable to find image 'oca-queue:fd5b8ac_env_1_job_2' locally
docker: Error response from daemon: pull access denied for oca-queue, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
```

Do you happen to know why? 😌

**Comment by guewen** [comment]
Oh maybe I'm looking at the wrong logs and the problem would be phantom-js.

**Comment by sbidoul** [comment]
Could it be the new(ish) docker hub rate limits that kick in ?

**Comment by StefanRijnhart** [comment]
I believe you can squash out the commit from #312 because that was a port forward of #311

File: queue_job/job.py
Comment (StefanRijnhart) [inline queue_job/job.py]: What do you think about undoing the changes from python-black in this PR? It would make it a lot easier to review.

File: queue_job/controllers/main.py
Comment (StefanRijnhart) [inline queue_job/controllers/main.py:147]: This is strange. In 13.0, this code is added by https://github.com/OCA/queue/pull/265 by @nilshamerlinck (and as such also present in https://github.com/OCA/queue/pull/330), but in your git history it is attributed to @guewen as part of https://github.com/OCA/queue/pull/333/commits/7385fd54fcbb2370413fb1fb305b6f213f8150a2.

I'm getting the feeling that maybe these backports should be split up in separate PRs.

Comment (etobella) [reply queue_job/controllers/main.py:147]: I think I did something wrong with the commits. I will recheck thanks :wink:

**Comment by etobella** [comment]
I think all the missing commits are here
Also, all the changes from black shouldn't be here.

**Comment by StefanRijnhart** [comment]
It's gotten bigger again, and do we really want to backport those large API changes to this older release?

**Comment by etobella** [comment]
It is bigger because I analyzed all the commits that where not backported. Originally, it was intended to backport them (maybe not all of them, but a lot), but it wasn't done.

**Comment by etobella** [comment]
@OCA/connector-maintainers 
The issue in runbot already exists on 12.0 branch of queue
https://runbot.odoo-community.org/runbot/repo/https-github-com-oca-queue-230

**Comment by guewen** [comment]
I tried to use this branch on a project and the migration failed, going to investigate why.

File: queue_job/migrations/12.0.2.3.0/post-migration.py
Comment (guewen) [inline queue_job/migrations/12.0.2.3.0/post-migration.py]: I had to do this: 

```suggestion
    exc_info = job.read(fields=['exc_info'])[0]['exc_info']
    for line in reversed(exc_info.splitlines()):
```

Otherwise, browsing the job could fail if the job is on a model which is not yet loaded by the registry:

```
Traceback (most recent call last):
  File "/odoo/src/odoo/service/server.py", line 1162, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/odoo/src/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/odoo/src/odoo/modules/loading.py", line 417, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/odoo/src/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/odoo/src/odoo/modules/loading.py", line 227, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "/odoo/src/odoo/modules/migration.py", line 190, in migrate_module
    migrate(self.cr, installed_version)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 14, in migrate
    _compute_jobs_new_values(env)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 21, in _compute_jobs_new_values
    exception_details = _get_exception_details(job)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 27, in _get_exception_details
    for line in reversed(job.exc_info.splitlines()):
  File "/odoo/src/odoo/fields.py", line 1036, in __get__
    self.determine_value(record)
  File "/odoo/src/odoo/fields.py", line 1139, in determine_value
    record._prefetch_field(self)
  File "/odoo/src/odoo/models.py", line 2880, in _prefetch_field
    result = records.read([f.name for f in fs], load='_classic_write')
  File "/odoo/src/odoo/models.py", line 2835, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/odoo/src/odoo/models.py", line 5167, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/odoo/src/odoo/fields.py", line 1044, in __get__
    return self.convert_to_record(value, record)
  File "/odoo/external-src/queue/queue_job/fields.py", line 64, in convert_to_record
    return json.loads(value or default, cls=JobDecoder, env=record.env)
  File "/usr/lib/python3.5/json/__init__.py", line 332, in loads
    return cls(**kw).decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/odoo/external-src/queue/queue_job/fields.py", line 107, in object_hook
    model = self.env[obj['model']]
  File "/odoo/src/odoo/api.py", line 831, in __getitem__
    return self.registry[model_name]._browse((), self)
  File "/odoo/src/odoo/modules/registry.py", line 177, in __getitem__
    return self.models[model_name]
KeyError: 'xxx.product.template'
```

**Comment by guewen** [comment]
Here's the correction for the issue I mentioned before https://github.com/OCA/queue/pull/333/files#r626409959

File: queue_job/migrations/12.0.2.3.0/post-migration.py
Comment (guewen) [reply queue_job/migrations/12.0.2.3.0/post-migration.py]: But we would better have to read all the "exc_info" at a time for all the jobs rather than 1 by 1 as I did here.

File: queue_job/migrations/12.0.2.3.0/post-migration.py
Comment (guewen) [inline queue_job/migrations/12.0.2.3.0/post-migration.py]: Better one reading all the exc_info at once:
```suggestion
def _compute_jobs_new_values(env):
    job_exc_infos = env["queue.job"].search(
        [("state", "=", "failed"), ("exc_info", "!=", False)]
    ).read(fields=["exc_info"])
    for job_row in job_exc_infos:
        exception_details = _get_exception_details(job_row["exc_info"])
        if exception_details:
            job = env["queue.job"].browse(job_row["id"])
            job.write(exception_details)


def _get_exception_details(exc_info):
    for line in reversed(exc_info.splitlines()):
        if _find_exception(line):
            name, msg = line.split(":", 1)
            return {
                "exc_name": name.strip(),
                "exc_message": msg.strip("()', \""),
            }
```

File: queue_job/migrations/12.0.2.3.0/post-migration.py
Comment (guewen) [reply queue_job/migrations/12.0.2.3.0/post-migration.py]: Here is the better proposal reading all the rows at once: https://github.com/OCA/queue/pull/333/files#r626439498

File: queue_job/migrations/12.0.2.3.0/post-migration.py
Comment (guewen) [reply queue_job/migrations/12.0.2.3.0/post-migration.py]: Hi @etobella, could you merge this suggestion?

Comment (etobella) [reply queue_job/migrations/12.0.2.3.0/post-migration.py]: Sure!!! Done!

Comment (guewen) [reply queue_job/migrations/12.0.2.3.0/post-migration.py]: Thanks :)

**Comment by guewen** [comment]
Hi @StefanRijnhart, do you still have concerns or can we merge the backports?

**Comment by StefanRijnhart** [comment]
For me this is too much to review at once, and to assess the impact of. But I do trust you fully with that, so don't mind me ;-)

File: queue_job/views/queue_job_views.xml
Comment (guewen) [inline queue_job/views/queue_job_views.xml]:
```suggestion
                <field name="exec_time" />
```

@etobella can you apply this change please? (I did the same for 13.0 in #352)

**Comment by guewen** [comment]
@etobella do you run this on production already?
FYI, as the company I work for is on 12.0 and is a large consumer of jobs, I should deploy this branch soon to put this to test before merging.

**Comment by etobella** [comment]
I did the backport in order to backport `edi`, but at the end we did not deploy the change. Sorry :cry:

**Comment by guewen** [comment]
> I did the backport in order to backport `edi`, but at the end we did not deploy the change. Sorry

No problem, I only asked to assess the risks ;)

**Comment by nilshamerlinck** [comment]
Hi, fyi I've been running this branch in production for 2 weeks without issue.

**Comment by guewen** [comment]
> Hi, fyi I've been running this branch in production for 2 weeks without issue.

Thanks for your feedback @nilshamerlinck!
Running well on the test (using production database) service here too.

Thanks @etobella for the backport :)

/ocabot merge major

**Comment by pedrobaeza** [comment]
Since the merge of this, I get continuous logs with this problem, and no job is executed:
<details>

```
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/serving.py", line 270, in run_wsgi
odoo_1    |     execute(self.server.app)               
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/serving.py", line 258, in execute
odoo_1    |     application_iter = app(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 142, in application
odoo_1    |     return application_unproxied(environ, start_response)                                                                                                                                
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 117, in application_unproxied
odoo_1    |     result = odoo.http.root(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1320, in __call__
odoo_1    |     return self.dispatch(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1293, in __call__
odoo_1    |     return self.app(environ, start_wrapped)
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/wsgi.py", line 766, in __call__
odoo_1    |     return self.app(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1488, in dispatch
odoo_1    |     result = ir_http._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/website_sale/models/ir_http.py", line 15, in _dispatch
odoo_1    |     return super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/auth_signup/models/ir_http.py", line 19, in _dispatch
odoo_1    |     return super(Http, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/web_editor/models/ir_http.py", line 22, in _dispatch
odoo_1    |     return super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/http_routing/models/ir_http.py", line 403, in _dispatch
odoo_1    |     result = super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/utm/models/ir_http.py", line 26, in _dispatch
odoo_1    |     response = super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 212, in _dispatch
odoo_1    |     return cls._handle_exception(e)
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/website/models/ir_http.py", line 195, in _handle_exception
odoo_1    |     return super(Http, cls)._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/utm/models/ir_http.py", line 31, in _handle_exception
odoo_1    |     response = super(IrHttp, cls)._handle_exception(exc)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 182, in _handle_exception
odoo_1    |     return request._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 776, in _handle_exception
odoo_1    |     return super(HttpRequest, self)._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 314, in _handle_exception
odoo_1    |     raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 87, in reraise
odoo_1    |     raise value
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 208, in _dispatch
odoo_1    |     result = request.dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 835, in dispatch
odoo_1    |     r = self._call_function(**self.params)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 346, in _call_function
odoo_1    |     return checked_call(self.db, *args, **kwargs)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 98, in wrapper
odoo_1    |     return f(dbname, *args, **kwargs)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
odoo_1    |     result = self.endpoint(*a, **kw)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 941, in __call__
odoo_1    |     return self.method(*args, **kw)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 519, in response_wrap
odoo_1    |     response = f(*args, **kw)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/controllers/main.py", line 81, in runjob
odoo_1    |     job = Job.load(env, job_uuid)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/job.py", line 252, in load
odoo_1    |     return cls._load_from_db_record(stored)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/job.py", line 263, in _load_from_db_record
odoo_1    |     method = getattr(recordset, method_name)
odoo_1    | AttributeError: 'base' object has no attribute 'confirm_one_invoice' - - -
```

</details>

Can you point which can be the problem?

**Comment by guewen** [comment]
@pedrobaeza can you paste the func_string, method_name, and records, args, kwargs json fields of one of these jobs?

**Comment by pedrobaeza** [comment]
Thanks for the quick reply. It's this function:

https://github.com/OCA/l10n-spain/blob/5c02c1d9caf61e02f57e2f2a4ec4c675283e92de/l10n_es_aeat_sii/models/account_invoice.py#L1450

and the job is this way:

Selección_030

This is before switching to new version.

Seeing 13.0 version, it seems to not be any difference:

https://github.com/OCA/l10n-spain/blob/23924f08218034c5f6b8f92ed2034e500070dd42/l10n_es_aeat_sii_oca/models/account_move.py#L1452

**Comment by StefanRijnhart** [comment]
Was this job created before you updated the code of the module and ran the module upgrade?

It looks like https://github.com/OCA/queue/pull/333/files#diff-1c5ee7d63e3224474074dbbdaa46e8d7a838199d2b4d4182726de22fd6e627d1R15-R16 might have been better as an end-migration script, because it might depend on the models and the methods being loaded at the time of migration.

If you query the database for the columns @guewen mentions, we can check this hypothesis.

-- edit -- that method looks safe though https://github.com/OCA/queue/blob/12.0/queue_job/models/queue_job.py#L563-L573

**Comment by pedrobaeza** [comment]
Yeah, it can be. For now, I have go back with the previous commit. As soon as I have a slot, I will propose a PR changing it to end-migration and try again to see if that's the problem.

**Comment by StefanRijnhart** [comment]
@pedrobaeza not sure that that's it. I have an environment with a lot of jobs on which I can check the migration, so I will post back about that as well.
BTW you might have to do some reverse migration if you want to go back, at least on jobs created after running with the code from this PR.
And it would still be informative to have the SQL values of the func_string, method_name, and records, args, kwargs columns of one of the failing jobs.

**Comment by guewen** [comment]
According to the screenshot above, the `func_string` is correct, so I don't think this migration is the problem.
According to the stack trace, the most likely is that this migration https://github.com/OCA/queue/blob/ad28e410e1d1c033f0df7810d086a134e210091d/queue_job/migrations/12.0.2.1.0/pre-migration.py#L20-L26 didn't run or for some reason generated a wrong json in the `records` field. The default value being the `base` model: https://github.com/OCA/queue/blob/ad28e410e1d1c033f0df7810d086a134e210091d/queue_job/fields.py#L33-L35

So:

1. Check if the migration 12.0.2.1.0 runs
2. Check and give feedback about the SQL content of the `records` field (func_string, method_name, args, kwargs may help but I really expect the issue to be in `records`).

**Pull request merged**
